### ISSUE TITLE
Resuming Training from 50k Checkpoint Leads to Loss Instability and reduced model performance

### ISSUE THREAD

**Issue by Tushar010K** [k2-fsa/OmniVoice#123]
### Checks

- [x] This template is only for usage issues encountered.
- [x] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [x] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [x] I am using English to submit this issue to facilitate community communication.

### Environment Details

OS: Ubuntu 
Python: 3.10.17
PyTorch: (installed with CUDA support)
CUDA: 12.4
GPU: NVIDIA GPU H200
Framework: accelerate

### Steps to Reproduce

1. Train the model on a dataset up to 50,000 steps
2. Save checkpoint (checkpoint-50000)
3. Modify dataset slightly (add new data)
4. Resume training using:
resume_from_checkpoint=/checkpoint-50000
5. Start training again

### ✔️ Expected Behavior

1. Training should resume seamlessly from step 50,000 
2. Model should continue learning without instability
3. Loss should remain stable or improve gradually
4. Dataloader should resume from previous state without repeated resets
5. Training should support newly added data without degrading performance

Configuration used 
{
    "llm_name_or_path": "Qwen/Qwen3-0.6B",
    "audio_vocab_size": 1025,
    "audio_mask_id": 1024,
    "num_audio_codebook": 8,

    "audio_codebook_weights": [8, 8, 6, 6, 4, 4, 2, 2],
    "drop_cond_ratio": 0.1,
    "prompt_ratio_range": [0.0, 0.3],
    "mask_ratio_range": [0.0, 1.0],
    "language_ratio": 0.8,
    "use_pinyin_ratio": 0.0,
    "instruct_ratio": 0.0,
    "only_instruct_ratio": 0.0,

    "resume_from_checkpoint": "/checkpoint-50000",
    "init_from_checkpoint": null,

    "learning_rate": 2e-5,
    "weight_decay": 0.01,
    "max_grad_norm": 1.0,
    "steps": 60000,
    "seed": 42,

    "warmup_type": "ratio",
    "warmup_ratio": 0.01,
    "warmup_steps": 0,

    "batch_tokens": 10288,
    "gradient_accumulation_steps": 1,
    "num_workers": 2,

    "mixed_precision": "no",
    "allow_tf32": true,

    "logging_steps": 50,
    "eval_steps": 2500,
    "save_steps": 2500,
    "keep_last_n_checkpoints": -1
}

### ❌ Actual Behavior

1. Behavior suggests possible mismatch in:
Dataloader state
Scheduler continuation
Dataset changes vs checkpoint state

**Comment by zhu-han** [comment]
When resuming training, the dataloader will start from scratch instead of resuming from its last state (because we are using an IterableDataset). As a result, the performance will differ from training entirely from scratch. I personally never used the resume training mode in my experiments. If you find this mode results in degraded performance, you can try fine-tuning using init_from_checkpoint.
